mutiny-core/src/auth.rs

```diff
@@ -0,0 +1,219 @@
+#![allow(dead_code)]
+use crate::{
+    error::MutinyError,
+    lnurlauth::{make_lnurl_auth_connection, AuthManager},
+    logging::MutinyLogger,
+    networking::websocket::{SimpleWebSocket, WebSocketImpl},
+    storage::MutinyStorage,
+};
+use jwt_compact::UntrustedToken;
+use lightning::util::logger::*;
+use lightning::{log_error, log_info};
+use lnurl::{lnurl::LnUrl, AsyncClient as LnUrlClient};
+use reqwest::Client;
+use reqwest::{Method, StatusCode, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct CustomClaims {
+    sub: String,
+}
+
+pub(crate) struct MutinyAuthClient<S: MutinyStorage> {
+    auth: AuthManager<S>,
+    lnurl_client: Arc<LnUrlClient>,
+    url: String,
+    http_client: Client,
+    jwt: RwLock<Option<String>>,
+    logger: Arc<MutinyLogger>,
+}
+
+impl<S: MutinyStorage> MutinyAuthClient<S> {
+    pub fn new(
+        auth: AuthManager<S>,
+        lnurl_client: Arc<LnUrlClient>,
+        logger: Arc<MutinyLogger>,
+        url: String,
+    ) -> Self {
+        let http_client = Client::new();
+        Self {
+            auth,
+            lnurl_client,
+            url,
+            http_client,
+            jwt: RwLock::new(None),
+            logger,
+        }
+    }
+
+    pub async fn authenticate(&self) -> Result<(), MutinyError> {
+        self.retrieve_new_jwt().await
+    }
+
+    pub fn is_authenticated(&self) -> bool {
+        if self.jwt.try_read().unwrap().is_some() {
+            return true; // TODO parse and make sure still valid
+        }
+        false
+    }
+
+    pub async fn request(
+        &self,
+        method: Method,
+        url: Url,
+        body: Option<Value>,
+    ) -> Result<reqwest::Response, MutinyError> {
+        let res = self
+            .authenticated_request(method.clone(), url.clone(), body.clone())
+            .await?;
+        match res.status() {
+            StatusCode::UNAUTHORIZED => {
+                // If we get a 401, refresh the JWT and try again
+                self.retrieve_new_jwt().await?;
+                self.authenticated_request(method, url, body).await
+            }
+            _ => Ok(res),
+        }
+    }
+
+    async fn authenticated_request(
+        &self,
+        method: Method,
+        url: Url,
+        body: Option<Value>,
+    ) -> Result<reqwest::Response, MutinyError> {
+        let mut request = self.http_client.request(method, url);
+
+        if self.jwt.try_read().unwrap().is_none() {
```

Comment:
should use `is_authenticated` here

Reply:
maybe should have it return `Option<String>` of the token and can use that for the request